src/_pytest/tmpdir.py

```diff
@@ -214,3 +229,25 @@ def tmp_path(request: FixtureRequest, tmp_path_factory: TempPathFactory) -> Path
     """
 
     return _mk_tmp(request, tmp_path_factory)
+
+
+def pytest_sessionfinish(session, exitstatus: Union[int, ExitCode]):
+    """After each session, remove base directory if all the tests passed,
+    the policy is "failed", and the basetemp is not specified by a user.
+    """
+    tmp_path_factory: TempPathFactory = session.config._tmp_path_factory
+    if tmp_path_factory._basetemp is None:
+        return
+    policy = tmp_path_factory._retention_policy
+    if (
+        exitstatus == 0
+        and policy == "failed"
+        and tmp_path_factory._given_basetemp is None
+    ):
+
+        def cleanup_passed_dir(passed_dir: Path):
+            if passed_dir.exists():
+                rmtree(passed_dir)
+
+        # Register to remove the base directory before starting cleanup_numbered_dir
+        atexit.register(cleanup_passed_dir, tmp_path_factory._basetemp)
```

Comment:
I would prefer to explicitly remove them at this point instead of relying on `atexit`.

pytest might be embedded, for example as part of user script or IDE, which might not actually exit the interpreter for a awhile.

Reply:
Done.
Is there any reason for using `atexit` to execute `cleanup_numbered_dir` in the first place? That makes it really difficult to test the behavior.